Add configuration workflow job, which automatically configures the rest of the workflow run

chrisjbillington opened the pull request:
And other changes aimed at minimising deployment and maintenance overhead. Ideally, one shouldn't need to update this workflow file for new Python versions, and should not need to edit it at all for each repository it is deployed to - the workflow file can be deployed as-is.

* No longer need to specify PURE or NOARCH - this will be introspected during the configure job.
* No longer need to specify the package name, it will be introspected during the configure job.
* No longer need to specify which versions of Python to build for in the case of a non-pure wheel or non-noarch conda package - the configure job will set variables to target all currently-supported Python versions.
* Use `cibuildwheel` to build impure wheels on all platforms instead of only `manylinux` for Linux (now deprecated).
* No longer need to adjust list of OSs to run jobs on - this will be determined automatically based on pure/noarch status.
* No longer need to specify the repository in job-level if-statements to prevent jobs running on forks. Instead, jobs simply do run on forks, but upload steps will be skipped if the relevant secrets are absent
* Removed "ignore tags" step that was designed to prevent running the workflow twice on a commit with tags and uploading twice to the Anaconda test label, resulting in an error. Instead, we upload releases to real PyPI and Anaconda main label,  and non-releases to test PyPI and anaconda test label so there is no duplication. Also we add the `--skip-existing` flag to anaconda uploads since sometimes you just need to re-run workflows anyway, this should be idempotent and not crash simply because some previous uploads succeeded.
* Add `if-no-files-found: error` to all `Upload Artifact` actions, with step-level `if` statements so they only run when we expect output. This reduces the incidence of failed runs that produced no artifacts appearing successful.
* Fix macOS impure conda builds. These previously didn't work at all on the newer macOS runners, because conda's compiler toolchains are super duper out of date. Instead, we use conda-forge toolchains by installing miniforge instead of miniconda.
* Use bash to build conda packages on Windows, same as Linux/macOS - it was never necessary to use cmd.exe in the first place, I think doing so just masked the path-length limit problem for some packages, which we now address by setting the conda build root path to a directory with a very short filepath.

#### Changes needed to use this workflow

It was not possible to get rid of absolutely all configuration, but at least we can obviate the need to edit the workflow file itself, which should now be able to be identical across repositories. To have this workflow do anaconda uploads, you'll need to set the `ANACONDA_USER` variable in your repository or organisation's Actions variables (similar to how secrets are set). I've already done this for the `labscript-suite` organisation.

chrisjbillington committed:
Add configuration workflow job, which automatically configures the rest of the workflow run

And other changes aimed at minimising deployment and maintenance
overhead. Ideally, one shouldn't need to update this workflow file for
new Python versions, and should not need to edit it at all for each
repository it is deployed to - the workflow file can be deployed
as-is.

* No longer need to specify PURE or NOARCH - this will be introspected
  during the configure job.
* No longer need to specify the package name, it will be introspected
  during the configure job.
* No longer need to specify which versions of Python to build for in the
  case of a non-pure wheel or non-noarch conda package - the configure
  job will set variables to target all currently-supported Python
  versions.
* Use `cibuildwheel` to build impure wheels on all platforms instead of
  only `manylinux` for Linux (now deprecated).
* No longer need to adjust list of OSs to run jobs on - this will be
  determined automatically based on pure/noarch status.
* No longer need to specify the repository in job-level if-statements to
  prevent jobs running on forks. Instead, jobs simply do run on forks,
  but upload steps will be skipped if the relevant secrets are absent
* Removed "ignore tags" step that was designed to prevent running the
  workflow twice on a commit with tags and uploading twice to the
  Anaconda test label, resulting in an error. Instead, we upload
  releases to real PyPI and Anaconda main label,  and non-releases to
  test PyPI and anaconda test label so there is no duplication. Also we
  add the `--skip-existing` flag to anaconda uploads since sometimes
  you just need to re-run workflows anyway, this should be idempotent
  and not crash simply because some previous uploads succeeded.
* Add `if-no-files-found: error` to all `Upload Artifact` actions, with
  step-level `if` statements so they only run when we expect output.
  This reduces the incidence of failed runs that produced no artifacts
  appearing successful.
* Fix macOS impure conda builds. These previously didn't work at all on
  the newer macOS runners, because conda's compiler toolchains are
  super duper out of date. Instead, we use conda-forge toolchains by
  installing miniforge instead of miniconda.
* Use bash to build conda packages on Windows, same as Linux/macOS - it
  was never necessary to use cmd.exe in the first place, I think doing
  so just masked the path-length limit problem for some packages, which
  we now address by setting the conda build root path to a directory
  with a very short filepath.

It was not possible to get rid of absolutely all configuration, but at
least we can obviate the need to edit the workflow file itself, which
should now be able to be identical across repositories. To have this
workflow do anaconda uploads, you'll need to set the `ANACONDA_USER`
variable in your repository or organisation's Actions variables
(similar to how secrets are set). I've already done this for the
`labscript-suite` organisation.

chrisjbillington committed:
Switch to `pyproject.toml`-only layout

hard-code release-branch-semver and no-local-version

These are no longer configured to be different in CI than locally

Drop minimum version requirements for build requirements where those
versions are now several years old

chrisjbillington commented:
Also switch to `pyproject.toml`-only layout and hard-code setuptools_scm version_schemes. The workflow no longer changes these in CI, it's easier to just always have `release-branch-semver` and `no-local-version` all the time on CI or otherwise.

dihm commented on .github/workflows/release.yml:
```diff
@@ -199,64 +259,65 @@ jobs:
       - name: Publish on TestPyPI
+        if: env.TESTPYPI_UPLOAD == 'true'
         uses: pypa/gh-action-pypi-publish@release/v1
         with:
           user: __token__
-          password: ${{ secrets.testpypi }}
+          password: ${{ secrets.TESTPYPI }}
           repository-url: https://test.pypi.org/legacy/
 
       - name: Publish on PyPI
-        if: contains(github.ref, '/tags')
+        if: env.PYPI_UPLOAD == 'true'
         uses: pypa/gh-action-pypi-publish@release/v1
         with:
           user: __token__
-          password: ${{ secrets.pypi }}
+          password: ${{ secrets.PYPI }}
```
Question and a request.

1) What is the appetite for switching to [trusted publisher](https://packaging.python.org/en/latest/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/) releasing (instead of API keys) for PyPI? In the spirit of simplifying the workflow, it is actually quite nice since API keys don't need to be configured. The only github thing that _should_ be done is configuring a named environment, and that is optional.

2) Strongly related to the above, could I request that the release job be broken up into separate jobs? I recently got bit by this workflow. Using trusted publisher releasing on testPyPI and PyPI combined with PEP 740 just getting supported causes annoying breakage. Problem is attestation files are built locally on upload and pollute the dist directory. If you release to testPyPI and PyPI in the same job, the files collide and break whichever one is second.

It has apparently been the recommendation of PyPA to use separate jobs (and named environments) for all release jobs for a while now, and they are [very adamant](https://github.com/pypa/gh-action-pypi-publish/issues/283#issuecomment-2499302876) that using a single job for multiple releases is Not Supported (TM).

chrisjbillington replied:
I've split the releases into multiple jobs - for consistency have split them all up.

I'm out of the loop with new publishing developments (have been vaguely ignoring prods and notifications I've seen pop up in my peripheral vision about it), but I'm keen to do whatever will delay bit-rot the longest and minimise the need to configure individual repositories (though it sounds like you need to do manual configuration on PyPI and Test PyPI per project still, oh well). But yeah if the existing method is deprecated it sounds like we should move to this trusted publisher release mechanism.

I don't know what attestations are, but are these a separate thing just happening behind the scenes now regardless, and we just need to split the jobs up to not mess with it?

Then there's digital signing of the GitHub release tarballs as well - that's documented in the trusted publisher releasing documentation, that sounds optional, and I'm not sure how it ties in with the other things or is totally unrelated.

> The only github thing that should be done is configuring a named environment, and that is optional.

Cool, so you're saying you could create the environment in the GitHub UI, or you could just do the following in the workflow as per the example in the docs?

```yaml
  publish-to-pypi:
    name: >-
      Publish Python 🐍 distribution 📦 to PyPI
    if: startsWith(github.ref, 'refs/tags/')  # only publish to PyPI on tag pushes
    needs:
    - build
    runs-on: ubuntu-latest
    environment:
      name: pypi
      url: https://pypi.org/p/<package-name>  # Replace <package-name> with your PyPI project name
    permissions:
      id-token: write  # IMPORTANT: mandatory for trusted publishing
```

If so that's fine - we'd insert `<package-name>` programmatically there from the introspected PKGNAME variable rather than modifying the workflow file per project (which I am EXTREMELY keen to avoid).


Let me know if I've misunderstood anything. Happy to give this a go.

dihm replied:
> Cool, so you're saying you could create the environment in the GitHub UI, or you could just do the following in the workflow as per the example in the docs?

I have been creating them through the UI, but reading some docs I think they are automatically created when the workflow is first run. Which is nice. One less manual thing to keep track of.

Overall, I've been fairly happy with the trusted publisher stuff. It is a little annoying to have to set it up for both PyPI and TestPyPI, but it's all on a single screen and all the inputs will basically be identical for each repo.

Also, I am super looking forward to the refreshed workflow. Keeping them functional has become a huge pain.

webknjaz replied:
> Cool, so you're saying you could create the environment in the GitHub UI, or you could just do the following in the workflow as per the example in the docs?

The environment is auto-created. But it's created without any useful configuration (like having scope secrets). One of my favorite uses, though, is the required reviewers feature — if you enable it, it's like adding a pause in your workflow run — it'll hold and only start the job with the environment once a human clicked a button. This is useful in that it gives one more control over the final pre-release check/gate and gives a chance to download the workflow-stored artifact and inspect them as much as you like.

If you don't configure this, the OIDC-based mechanism in Trusted Publishing will still use the env name to match the expectations you've set up on the PyPI side (the name should be the same).


> If so that's fine - we'd insert `<package-name>` programmatically there from the introspected PKGNAME variable rather than modifying the workflow file per project (which I am EXTREMELY keen to avoid).

Moreover, you can also compute the version-specific page. These links will be preserved on the deployments page. Example: https://github.com/cherrypy/cheroot/deployments/pypi. URL rendering: https://github.com/cherrypy/cheroot/actions/runs/8786461123/workflow#L1181.

chrisjbillington committed:
Split release/upload steps into separate jobs

This is important for moving to PyPI attested uploads and is generally
recommended by the PyPI upload actions anyway.

We split the other steps into separate jobs for consistency

webknjaz commented on pyproject.toml:
```diff
@@ -1,7 +1,66 @@
 [build-system]
-requires = ["setuptools>=42", "wheel", "setuptools_scm>=4.1.0"]
+requires = ["setuptools", "wheel", "setuptools_scm"]
```
`wheel` was in the `setuptools`' docs examples for some while, but was never actually needed. Old setuptools versions would auto-inject the dep, and the modern ones vendor it. It shouldn't be listed, ever.
```suggestion
requires = ["setuptools", "setuptools_scm"]
```

chrisjbillington replied:
Ah, great to know, thanks!

chrisjbillington committed:
Trusted Publisher uploads to PyPI and TestPyPI

Since we can no longer use the presence of TestPyPI and PyPI API keys to
know if we should attempt to upload releases (i.e. we're not running in
a fork), we instead do need some explicit configuration.

Rather than do this all in repository variables, we do it in a config
file in the workflow directory.

RELEASE_REPO is set to the repository releases should be made from
(otherwise uploads are skipped).

For consistency, ANACONDA_USER is set here too rather than as a
repository variable.

For completeness everything else is configurable there too.

Re-add the "ignore tags" step - with fixes since it was incorrect given
recent changes to the workflow `on:` configuration. It is needed due to
race conditions - we are cloning the repo in multiple jobs and a tag
could appear in between them, leading to version inconsistencies.

chrisjbillington committed:
Enable verbose output for PyPI uploads and skipping existing distribution files

Particularly with TestPyPI uploads, the version number can change in
unexpected ways when tags are added after the fact, allowing for
collisions. Just skip in this case.

For PyPI uploads we don't expect collisions, but still allow skipping
instead of erroring in case e.g. some uploads fail and we need to retry
a workflow job

chrisjbillington commented:
Ok, this is working, and I've used it in `qtutils` (https://github.com/philipstarkey/qtutils/actions/runs/12364591645) and `zprocess` (https://github.com/chrisjbillington/zprocess/actions/runs/12365789345).

To enable trusted publisher releases, one just needs to go to:

https://pypi.org/manage/project/qtutils/settings/publishing/

(or the same with test.pypi) and fill in the details. We *are* using the optional environments (named `pypi` and `testpypi`).

Unfortunately if we're using trusted publisher releases, then we can't infer whether we should attempt to upload to PyPI/testPyPI or not based on the presence of the needed secret. So we do need some per-repository configuration after all. Rather than make this either in the form of repository variables, or hard-coded within the workflow file itself, I've decided on having a separate `release-vars.sh` that lives next to `release.yml` in the repository and contains per-repository configuration. It's a bash script that is sourced by the configuration job.

I made most things configurable even though everything can be introspected other than `RELEASE_REPO` (the repository from which package uploads should be made, otherwise they're skipped), and `ANACONDA_USER`.

So for deployment to other repos, we'll need to copy both files into other repos and modify `release-vars.sh` with the name of each repository.

This is simple enough that it could be scripted to roll it out to master and release branches on all repos, though I'd probably make a chacklist of other changes to make at the same time, such as:

* explicitly depending on `setuptools_scm` in runtime dependencies — necessary for `__version__` to work in editable installs
* removing `wheel` from build-system/requires
* removing `importlib_metadata` from dependencies and depending on python≥3.8 from which it's in the stdlib (3.7 and lower are end of life so this is kosher)
* archiving all but the latest "maintenance" branches
* simplifying declared Python version compatibility to just say "3" so we don't have to update it constantly - we aim for compatibility with all stable Python versions, no need to be more specific.

I'll merge this soon if there aren't objections or additional suggestions. Have already done the trusted publisher setup for this repo.

dihm commented:
Well I think this is a pretty solid upgrade. Keeping the config options localized is a decent compromise.

As a suggestion (that you have probably already considered), could we introspect if the repo running the workflow is a fork so we don't have to set the RELEASE_REPO? A quick look showed this SO suggestion (with comments being important as well). 

https://stackoverflow.com/questions/70775445/determining-the-origin-repo-of-pull-request-in-workflow-yaml

Even if that works, I still like the idea of a shell config script that collects all the important things (selecting release targets etc), even if the defaults are what we want most of the time.

dihm commented:
Note: for a non-pure installation you get an error configuring on apparently any `ci-helper distinfo` command if the `setup.py` is not importable. This came up when updating the `labscript-c-extensions` [workflow](https://github.com/labscript-suite/labscript-c-extensions/actions/runs/14407740179) which imports cython in `setup.py`.

chrisjbillington committed:
Update macos intel runner and miniforge args

* macos-13 → macos-15-intel as former is retired
* new syntax for auto activate based environment in setup-minoforge per
  deprecation warning

dihm commented:
> @dihm I've dealt with this in labscript-c-extensions by moving the cython extension details into pyproject.toml, which setuptools recently added support for:
> 
> [labscript-suite/labscript-c-extensions#7](https://github.com/labscript-suite/labscript-c-extensions/pull/7)
> 
> Merging this, as the rot continues and dealing with it should be done on top of this

That is cool about cython extensions! Now if only they could support dynamic linking to numpy...
